File: db/nodejs/src/core/segment-manager.ts
import {promises as fs} from "node:fs";
import {FileAdapter} from "./file.adapter";

export class SegmentManager {
    private currentSegmentNumber: number;

    constructor(protected fileAdapter: FileAdapter, protected segmentSize = 1024) {
        // this.baseFilename = resolve(baseFilename);
        this.segmentSize = segmentSize;
        this.initCurrentSegment();
        // this.currentFile = this.getSegmentFilename(this.currentSegment);
    }

    initCurrentSegment() {
        this.currentSegmentNumber = 0;
    }

    async set(key: string, data: any) {
        const size = await this.fileAdapter.getFileSize(this.currentSegmentNumber)
        if (size + Buffer.byteLength(JSON.stringify(data), 'utf-8') > this.segmentSize) {
            this.currentSegmentNumber++;
        }

        const result = await this.fileAdapter.set(key, data, this.currentSegmentNumber);
        return {offset: result.offset, segmentNumber: this.currentSegmentNumber};
    }

    async get(offset: number, segmentNumber: number) {
        return await this.fileAdapter.get(offset, segmentNumber);
    }

    // getSegmentFilename(segment) {
    //     return `${this.baseFilename}_segment_${segment}.txt`;
    // }
    //
    // async appendEntry(entry) {
    //     const fileHandle = await fs.open(this.currentFile, 'a');
    //     const offset = (await fileHandle.stat()).size;
    //
    //     if (offset + Buffer.byteLength(entry, 'utf-8') > this.segmentSize) {
    //         this.currentSegment++;
    //         this.currentFile = this.getSegmentFilename(this.currentSegment);
    //     }
    //
    //     await fileHandle.appendFile(entry);
    //     await fileHandle.close();
    //
    //     return { segment: this.currentSegment, offset };
    // }
    //
    // async readEntry(segment, offset) {
    //     const segmentFile = this.getSegmentFilename(segment);
    //     const fileHandle = await fs.open(segmentFile, 'r');
    //
    //     const buffer = Buffer.alloc(1024);
    //     await fileHandle.read(buffer, 0, 1024, offset);
    //     const line = buffer.toString('utf-8').trim();
    //
    //     await fileHandle.close();
    //
    //     return line;
    // }
    //
    // async rebuildIndex(index) {
    //     for (let i = 0; i <= this.currentSegment; i++) {
    //         const segmentFile = this.getSegmentFilename(i);
    //         const fileStream = createReadStream(segmentFile, 'utf-8');
    //         const rl = readline.createInterface({
    //             input: fileStream,
    //             crlfDelay: Infinity
    //         });
    //
    //         let offset = 0;
    //
    //         for await (const line of rl) {
    //             const [key] = line.split(',');
    //
    //             index.set(key, { segment: i, offset });
    //
    //             offset += Buffer.byteLength(line, 'utf-8') + 1;
    //         }
    //     }
    // }
}

File: db/golang/v1/internal/segment_manager/segment_manager.go
package segmentmanager

import (
	"fmt"

	fa "samurai-db/internal/file_adapter"
)

type SegmentManager struct {
	currentSegment int64
	segmentSize    int64
	fileAdapter    *fa.FileAdapter
}

func NewSegmentManager(fa *fa.FileAdapter) *SegmentManager {
	return &SegmentManager{fileAdapter: fa, currentSegment: 0, segmentSize: 1024}
}

func (s *SegmentManager) Set(key string, data any) (int64, int64, error) {
	foundCurrentSegmentFolder := true

	// fixme: Костыльный поиск последнего сегмента
	for foundCurrentSegmentFolder {
		size, _ := s.fileAdapter.GetFileSize(s.currentSegment)

		entry := s.fileAdapter.StringifyEntry(key, data)

		if size+int64(len(entry)) > s.segmentSize {
			s.currentSegment++
			continue
		}

		foundCurrentSegmentFolder = false
	}

	offset, err := s.fileAdapter.Set(key, data, s.currentSegment)
	if err != nil {
		return 0, 0, fmt.Errorf("set in segment manager: %e", err)
	}

	return offset, s.currentSegment, nil
}

func (s *SegmentManager) Get(offset, segment int64) (map[string]any, error) {

	return s.fileAdapter.Get(offset, segment)
}
